---
title: "DAAD RNA-seq course - lessons 3-4"
author: "Serhiy Naumenko"
date: "`r Sys.Date()`"
output:
  pdf_document:
    toc: yes
  html_document:
    code_folding: hide
    df_print: paged
    highlights: pygments
    number_sections: yes
    self_contained: yes
    theme: default
    toc: yes
    toc_float:
      collapsed: yes
      smooth_scroll: yes
---

# Overview
- Petrenko2024 RNA-seq experiment for reanalysis


```{r include = FALSE}
### Bioconductor and CRAN libraries used
library(tidyverse)
library(readr)
library(tidyr)
library(stringr)
# install.packages("BiocManager")
# BiocManager::install("AnnotationHub")
library(AnnotationHub) # bioc
library(ensembldb) # bioc
library(RColorBrewer)

library(DESeq2) # bioc
library(pheatmap) # R
library(DEGreport) # bioc
library(tximport) # bioc
library(ggplot2) 
library(ggrepel) #r
library(knitr)

library(ggplotify)#r
library(writexl) #r

library(clusterProfiler) #bioc
library(org.Mm.eg.db) # bioc

ggplot2::theme_set(theme_light(base_size = 14))

opts_chunk[["set"]](
    cache = FALSE,
    dev = c("png", "pdf"),
    error = TRUE,
    highlight = TRUE,
    message = FALSE,
    prompt = FALSE,
    tidy = FALSE,
    warning = FALSE)

```

\newpage

# Load annotations
```{r include = TRUE}
annotation_file <- "../../99_technical/ensembl112_mm10_annotations.txt"
gene_symbol_file <- "../../99_technical/gene_symbol.txt"
if (file.exists(annotation_file)){
    mmdb <- read_tsv(annotation_file)
    gene_symbol <- read_tsv(gene_symbol_file)
}else{
    # Connect to AnnotationHub
    ah <- AnnotationHub()
    # Query AnnotationHub
    mm_ens <- query(ah, c("Mus musculus", "EnsDb"))

    # Get Ensembl 112
    # AH116909
    mm_ens <- mm_ens[["AH116909"]]

    # Extract gene-level information
    txdb <- transcripts(mm_ens,
                    return.type = "data.frame") %>% 
    dplyr::select(tx_id, gene_id)  

    genedb <- genes(mm_ens, 
              return.type = "data.frame") %>% 
    dplyr::select(gene_id, gene_name, symbol)  

    gene_symbol <- genedb %>% dplyr::select(gene_id, symbol)
    write_tsv(gene_symbol, gene_symbol_file)

    mmdb <- inner_join(txdb, genedb)
    write.table(mmdb, 
            file = annotation_file , 
            sep = "\t", 
            row.names = F, 
            quote = F)
}
tx2gene <- mmdb[, c("tx_id", "gene_id")]
```

\newpage

# Load Counts
```{r}
# raw counts downloaded from 
# https://www.ebi.ac.uk/biostudies/arrayexpress/studies/E-MTAB-13804

# setwd("02_classes/03_rnaseq_intro_part1/")

protein_coding_genes <- read_csv("../../99_technical/ensembl_w_description.mouse.protein_coding.csv")

counts_csv <- "../../01_data/counts.csv"
counts_tpm_csv <- "../../01_data/count_matrix_tpm.csv"
if (file.exists(counts_csv)){
    counts_prepared <- read_csv(counts_csv)
    counts_tpm <- read_csv(counts_tpm_csv)
}else{
    counts_raw_csv <- "../../01_data/count_matrix_raw.csv"
    counts_raw <- read_csv(counts_raw_csv) 
    colnames(counts_raw)[1] <- "gene_name"
    # counts_tpm <- read_csv(counts_tpm_csv)
    # use Ensembl_Gene_id
    # remove genes with NA
    # filter protein coding genes

    counts_prepared <- counts_raw %>% left_join(gene_symbol, by = c("gene_name" = "symbol")) %>%
       dplyr::select(-gene_name) %>% drop_na(gene_id) %>% 
       semi_join(protein_coding_genes, by = c("gene_id" = "ensembl_gene_id")) %>% 
       relocate(gene_id) %>% rename(ensembl_gene_id = gene_id)
       
    gene_length <- read_tsv("../../99_technical/GC_lengths.tsv") 
    counts <- counts_prepared %>% arrange(ensembl_gene_id)
    gene_ids <- intersect(counts$ensembl_gene_id, gene_length$ensembl_gene_id)
    
    v_len <- gene_length %>% dplyr::filter(ensembl_gene_id %in% gene_ids)
    counts_prepared <- counts %>% dplyr::filter(ensembl_gene_id %in% gene_ids)

    write_csv(counts_prepared, counts_csv)
    counts <- counts %>% column_to_rownames("ensembl_gene_id")
    
    x <- counts / v_len$Length
    counts_tpm <- t(t(x) * 1e6 / colSums(x)) %>% as.data.frame() %>% round(2) %>% 
      rownames_to_column("ensembl_gene_id") %>% left_join(gene_symbol, 
                                                          by = c("ensembl_gene_id" = "gene_id")) %>%
      write_csv(counts_tpm_csv)
    
}
counts <- counts_prepared %>% column_to_rownames(var = "ensembl_gene_id")
```

\newpage

# Cleanup and load metadata
```{r}
# Load the data and metadata
# remove duplicate rows
metadata_csv <- "../../01_data/metadata.csv"
if (file.exists(metadata_csv)){
     metadata <- read_csv(metadata_csv)
}else{
    metadata_raw <- read_tsv("../../01_data/E-MTAB-13804.sdrf.txt") %>%
                      dplyr::select(-any_of(c("Scan Name", "Comment[SUBMITTED_FILE_NAME]", 
                            "Comment[ENA_RUN]", "Comment[FASTQ_URI]"))) %>% distinct() %>% 
                            dplyr::rename(sample_id = `Source Name`) 
    colnames(metadata_raw)[7] <- "age"
    colnames(metadata_raw)[44] <- "stimulus1"
    colnames(metadata_raw)[46] <- "factor_value"
    
    
    metadata_raw$stimulus <- str_replace_all(metadata_raw$stimulus1, "control \\(olive oil\\)", "olive_oil")
    metadata_raw$stimulus <- str_replace_all(metadata_raw$stimulus, "control \\(saline\\)", "saline")
    metadata_raw$stimulus <- str_replace_all(metadata_raw$stimulus, "carbon tetrachloride", "carbon_tetrachloride")
    
    metadata <- metadata_raw %>% separate(factor_value, sep = "_", 
                                          into = c("experiment", "treatment")) %>%
         dplyr::select(sample_id, experiment, treatment, stimulus)
    
    write_csv(metadata, metadata_csv)
}

metadata <- metadata %>% column_to_rownames(var = "sample_id") 
```

\newpage

# Run DESeq2
```
estimating size factors
estimating dispersions
gene-wise dispersion estimates
mean-dispersion relationship
final dispersion estimates
fitting model and testing
```

* [Estimating size factors and count normalization](https://hbctraining.github.io/DGE_workshop_salmon/lessons/02_DGE_count_normalization.html)
* [Gene-wise dispersions](https://hbctraining.github.io/DGE_workshop_salmon/lessons/04_DGE_DESeq2_analysis.html)
* [Mean-dispersion(variance) relationship and the Negative Binomial Model](https://hbctraining.github.io/DGE_workshop_salmon/lessons/01_DGE_setup_and_overview.html#modeling-count-data)
* [Model fitting and hypothesis testing](https://hbctraining.github.io/DGE_workshop_salmon/lessons/05_DGE_DESeq2_analysis2.html)


```{r include = TRUE}
## Create DESeq2Dataset object
dds_file <- "../../data/dds.RDS"

if (file.exists(dds_file)){
    dds <- readRDS(dds_file)
}else{
    dds <- DESeqDataSetFromMatrix(countData = counts, 
                              colData = metadata, 
                              design = ~treatment)

    
    # subset protein-coding genes
    pc_genes <- intersect(protein_coding_genes$ensembl_gene_id, row.names(dds))
    dds <- dds[pc_genes,]
    # 9509 genes left
    keep <- rowMeans(counts(dds)) >= 100
    dds <- dds[keep, ]

    # Run DESeq2
    dds <- DESeq(dds)
    saveRDS(dds, dds_file)
}
```


\newpage

# Convenience functions

```{r}

# return mean counts for a group of sample in a column
get_counts_for_samples <- function(ctpm, samples, column_name){
    tpm_counts <- ctpm %>%
        column_to_rownames("ensembl_gene_id") %>% 
        dplyr::select(any_of(samples)) %>% 
        rowMeans() %>%
        as.data.frame() %>%
        round(2) %>%
        rownames_to_column("ensembl_gene_id") 
    
    colnames(tpm_counts) <- c("ensembl_gene_id", "tpm")
    
    tpm_counts <- tpm_counts %>% 
        dplyr::mutate("{column_name}" := round(tpm, 2)) %>% 
        dplyr::select(-tpm)
    
    return(tpm_counts)
}

# get rid of excess precision
comb_de_result_table <- function(results){
    results <- results %>%
        mutate(baseMean = round(baseMean, 2),
               log2FoldChange = round(log2FoldChange, 2),
               lfcSE = round(lfcSE, 2),
               stat = round(stat, 2),
               pvalue = format(pvalue, scientific = TRUE, digits = 2),
               padj = format(padj, scientific = TRUE, digits = 2))
    return(results)
}

```

\newpage

# Sample-level QC analysis 
```{r include = FALSE}
### Transform counts for data visualization (unsupervised analysis)
rld_file <- "../../data/rld.RDS"
if (file.exists(rld_file)){
    rld <- readRDS(rld_file)
}else{
    #rld <- rlog(dds, blind = TRUE)
    rld <- vst(dds)
    saveRDS(rld, rld_file)
}
class(rld) # what type of object is this

# we also need just a matrix of transformed counts
rld_mat <- assay(rld)
```

# PCA - treatment
```{r, fig.width = 10, fig.height = 10}
plotPCA(rld, intgroup = c("treatment")) + 
  geom_label_repel(aes(label = name)) + 
  theme_bw()
```
\newpage

# PCA - experiment
```{r, fig.width = 10, fig.height = 10}
plotPCA(rld, intgroup = c("experiment")) + geom_label_repel(aes(label = name)) + theme_bw()
```

\newpage

# Clustering using top 1000 variable genes
```{r, fig.width = 20, fig.height = 20}
rv <- rowVars(rld_mat)
rv <- order(rv, decreasing = TRUE) %>% head(1000)
rld_mat_1000 <- rld_mat[rv,]
annotation <- metadata[, c("experiment", "treatment")]

# Change colors
heat.colors <- brewer.pal(6, "Blues")
rld_cor <- cor(rld_mat_1000)
# Plot heatmap
pheatmap(rld_cor, 
         annotation = annotation, 
         border = NA,
         fontsize = 20)
```



\newpage
# PCA: Controls
```{r, fig.width = 10, wig.height = 10}
rld.sub <- rld[ , rld$stimulus %in% c("saline", "olive_oil") ]
plotPCA(rld.sub, intgroup = c("stimulus")) + 
  geom_label_repel(aes(label = name)) + 
  theme_bw()

```

\newpage
# PCA: TAA and CCL4
```{r, fig.width = 10, fig.height = 10}
rld.sub <- rld[ , rld$stimulus %in% c("carbon_tetrachloride", "thioacetamide")]
plotPCA(rld.sub, intgroup = c("stimulus")) + 
  geom_label_repel(aes(label = name)) + 
  theme_bw()

```

\newpage
# PCA: TAA and CCL4 - treatment
```{r, fig.width = 10, fig.height = 10}
rld.sub <- rld[ , rld$stimulus %in% c("carbon_tetrachloride", "thioacetamide")]
plotPCA(rld.sub, intgroup = c("treatment", "stimulus")) + 
  geom_label_repel(aes(label = name)) + 
  theme_bw()

```

\newpage
# PCA: TAA experiment
```{r, fig.width = 12, fig.height = 12}
rld.sub <- rld[ , rld$experiment %in% c("TAA")]
plotPCA(rld.sub, intgroup = c("treatment")) + 
  geom_label_repel(aes(label = name)) + theme_bw(base_size = 15)

```


\newpage
# PCA: CCL4 experiment
```{r, fig.width = 10, fig.height = 10}
rld.sub <- rld[ , rld$experiment %in% c("CCL4")]
plotPCA(rld.sub, intgroup = c("treatment")) + 
  geom_label_repel(aes(label = name)) + 
  theme_bw()

```


\newpage
# DE in TAA
```{r}
ddsTAA <- subset(dds, select = colData(dds)$experiment == "TAA")
ddsTAA <- subset(ddsTAA, select = colData(ddsTAA)$treatment == "HC" | colData(ddsTAA)$treatment == "CIR")
ddsTAA$treatment <- droplevels(ddsTAA$treatment)
ddsTAA$treatment <- relevel(ddsTAA$treatment, ref = "HC")

contrast <- c("treatment", "CIR", "HC")
ddsTAA <- DESeq(ddsTAA)

resTreatment <- results(dds, contrast = contrast, alpha = 0.05)
length(which(resTreatment$padj < 0.05))

# Add annotations
resTreatment_tb <- resTreatment %>%
  data.frame() %>% 
  rownames_to_column(var = "gene") %>% 
  as_tibble() %>% 
  left_join(gene_symbol, by = c("gene" = "gene_id"))

resTreatment_tb_significant <- dplyr::filter(resTreatment_tb, padj < 0.05) %>%  
                        dplyr::filter(abs(log2FoldChange) > 1) %>% 
                        comb_de_result_table()

write_csv(resTreatment_tb_significant, "../../03_outputs/T4.TAA_results.csv")

samples_control <- metadata %>% rownames_to_column("sample") %>% 
         dplyr::filter(experiment == "TAA" & treatment == "HC")  %>% pull(sample)

counts_tpm$symbol <- NULL
tpm_control <- get_counts_for_samples(counts_tpm, samples_control, "HC_mean_tpm")

samples_effect <- metadata %>% dplyr::filter(experiment == "TAA" & treatment == "CIR") %>% row.names() 
tpm_effect <- get_counts_for_samples(counts_tpm, samples_effect, "CIR_tpm")

tpm_counts <- tpm_effect %>% 
              left_join(tpm_control,
                        by = c("ensembl_gene_id" = "ensembl_gene_id"))

resTreatment_tb_significant <- resTreatment_tb_significant %>% 
          left_join(tpm_counts, by = c("gene" = "ensembl_gene_id")) %>%
          arrange(log2FoldChange)

write_xlsx(list(T4.TAA_results = resTreatment_tb_significant), 
          "../../03_outputs/T4.DE_TAA.xlsx")

# Separate into up and down-regulated gene sets
sigTreatment_up <- rownames(resTreatment)[which(resTreatment$padj < 0.01 & resTreatment$log2FoldChange > 0)]
sigTreatment_down <- rownames(resTreatment)[which(resTreatment$padj < 0.01 & resTreatment$log2FoldChange < 0)]
```



\newpage

# Visualization - Gene example

```{r, fig.width=12, fig.height = 12}
d <- plotCounts(dds, 
                gene = "ENSMUSG00000027204", 
                intgroup = "treatment", 
                returnData = TRUE)

ggplot(d, aes(x = treatment, y = count, color = treatment)) + 
     geom_point(position = position_jitter(w = 0.1, h = 0)) +
     geom_text_repel(aes(label = rownames(d))) + 
     theme_bw(base_size = 10) +
     ggtitle("Fbn1") +
     theme(plot.title = element_text(hjust = 0.5)) +
     scale_y_log10()
     
```


\newpage
# Heatmaps
```{r, fig.width = 20, fig.height = 30}
# Create a matrix of normalized expression
sig_up <- resTreatment_tb_significant %>% arrange(-log2FoldChange) %>% head(50) %>% pull(gene)
sig_down <- resTreatment_tb_significant %>% arrange(log2FoldChange) %>% head(50) %>% pull(gene)
sig <- c(sig_up, sig_down)

row_annotation <- gene_symbol %>% 
                    as_tibble() %>% 
                    dplyr::filter(gene_id %in% sig)

plotmat <- counts_tpm %>% column_to_rownames("ensembl_gene_id") %>% 
     dplyr::select(any_of(c(samples_control, samples_effect)))
                                                                                  
plotmat <- plotmat[c(sig_up, sig_down),] %>% as.data.frame() %>% 
          rownames_to_column(var = "ensembl_gene_id") %>% 
          left_join(gene_symbol, by = c("ensembl_gene_id" = "gene_id")) %>% 
          drop_na(symbol)

plotmat$ensembl_gene_id <- NULL

plotmat <- plotmat %>% column_to_rownames(var = "symbol") %>% as.matrix()

# Color palette
heat.colors <- brewer.pal(6, "YlOrRd")

# Plot heatmap
# color = heat.colors,
pheatmap(plotmat, 
         scale = "row", 
         show_rownames = TRUE,
         border = FALSE,
         annotation = metadata[, c("treatment"), drop = FALSE],
         main = "Top 50 Up- and Down- regulated genes in treatment: CIR vs HC",
         fontsize = 20)

```

# Functional analysis

## Create background dataset for hypergeometric testing using all genes tested for significance in the results                 
```{r, fig.width=10, fig.height=10}
bg_genes <- resTreatment_tb$gene

## Run GO enrichment analysis 
compGO <- enrichGO(gene = sigTreatment_up,
                            universe = bg_genes,
                            keyType = "ENSEMBL",
                            OrgDb = "org.Mm.eg.db", 
                            ont = "BP", 
                            qvalueCutoff  = 0.05, 
                            pAdjustMethod = "BH")

dotplot(compGO,
        showCategory = 20, 
        title = "GO (Biological Process) Enrichment Analysis for UP in Treatment")
# image pdf 12 x 12

```


```{r, fig.width=10, fig.height=10}
## Run GO enrichment analysis 
compGO <- enrichGO(gene = sigTreatment_down,
                            universe = bg_genes,
                            keyType = "ENSEMBL",
                            OrgDb = org.Mm.eg.db, 
                            ont = "BP", 
                            qvalueCutoff  = 0.05, 
                            pAdjustMethod = "BH")

dotplot(compGO, showCategory = 25, title = "GO (Biological Process) Enrichment Analysis for DOWN in Treatment")
```

\newpage

# R session
```{r}
sessionInfo()
```